### PR TITLE
feat: kafka topic provisoner

### PR DESCRIPTION
**Pull request by chrisgacsal**
## Overview

* add `TopicProvisioner` for creating/deleting Kafka topics in a unified way
* replace different Kafka topic provisioning methods with `TopicProvisioner`

### CODE REVIEW

**Review by hekike — APPROVED** [review]

File: cmd/notification-service/main.go [hunk 1]
```diff
@@ -399,18 +408,42 @@ func wmBrokerConfiguration(conf config.Configuration, logger *slog.Logger, metri
...
-func initEventPublisherDriver(ctx context.Context, logger *slog.Logger, conf config.Configuration, metricMeter metric.Meter) (message.Publisher, error) {
-	provisionTopics := []watermillkafka.AutoProvisionTopic{}
-	if conf.Notification.Consumer.DLQ.AutoProvision.Enabled {
-		provisionTopics = append(provisionTopics, watermillkafka.AutoProvisionTopic{
-			Topic:         conf.Notification.Consumer.DLQ.Topic,
-			NumPartitions: int32(conf.Notification.Consumer.DLQ.AutoProvision.Partitions),
-			Retention:     conf.BalanceWorker.DLQ.AutoProvision.Retention,
+func initEventPublisherDriver(ctx context.Context, logger *slog.Logger, conf config.Configuration, metricMeter metric.Meter, topicProvisioner pkgkafka.TopicProvisioner) (message.Publisher, error) {
+	var provisionTopics []pkgkafka.TopicConfig
+	if conf.BalanceWorker.DLQ.AutoProvision.Enabled {
```
Comment (sagikazarmark) [inline cmd/notification-service/main.go:413]: This is the notification service, isn't it?

File: config/ingest.go [hunk 2]
```diff
@@ -14,6 +14,12 @@ import (
 
 type IngestConfiguration struct {
 	Kafka KafkaIngestConfiguration
+
+	// The maximum number of entries stored in topic cache at a time which after the least recently used is evicted.
+	// Setting size to 0 makes it unlimited
+	CacheSize int
```
Comment (sagikazarmark) [inline config/ingest.go:20]:
```suggestion
	CacheSize int

```

File: pkg/kafka/topicprovisioner.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+type TopicProvisionerConfig struct {
+	AdminClient *kafka.AdminClient
+	Logger      *slog.Logger
+	Meter       metric.Meter
+
+	// CacheSize stores he maximum number of entries stored in topic cache at a time which after the least recently used is evicted.
+	// Setting it to 0 makes the cache size unlimited.
+	CacheSize int
```
Comment (sagikazarmark) [inline pkg/kafka/topicprovisioner.go:47]:
```suggestion
	CacheSize int

```

File: cmd/notification-service/main.go (hunk 1 above)
Comment (chrisgacsal) [reply cmd/notification-service/main.go:413]: Nice catch! Will fix it.

File: cmd/balance-worker/main.go [hunk 4]
```diff
@@ -370,18 +379,42 @@ func wmBrokerConfiguration(conf config.Configuration, logger *slog.Logger, metri
...
-		ProvisionTopics: provisionTopics,
+		Broker:           broker,
+		ProvisionTopics:  provisionTopics,
+		TopicProvisioner: topicProvisioner,
+	})
+}
+
+func initTopicProvisioner(conf config.Configuration, logger *slog.Logger, meter metric.Meter) (pkgkafka.TopicProvisioner, error) {
+	kafkaConfigMap := conf.Ingest.Kafka.CreateKafkaConfig()
+	delete(kafkaConfigMap, "go.logs.channel.enable")
```
[the remark this answers is not in the document]
Comment (chrisgacsal) [reply cmd/balance-worker/main.go:401]: Will do.

[the remark this answers is not in the document]
Comment (chrisgacsal) [reply cmd/balance-worker/main.go:403]: Done.

File: config/ingest.go (hunk 2 above)
Comment (chrisgacsal) [reply config/ingest.go:20]: Done.

File: pkg/kafka/topicprovisioner.go (hunk 3 above)
Comment (chrisgacsal) [reply pkg/kafka/topicprovisioner.go:47]: Done.

**Review by sagikazarmark — APPROVED** [review]

**Pull request merged**